### PR TITLE
[java] LocalVariableCouldBeFinal: allow excluding the variable in a for-each loop

### LINKED ISSUE #1513: [java] LocalVariableCouldBeFinal: allow excluding the variable in a for-each loop
Currently LocalVariableCouldBeFinal also triggers on the non-final variable in a for-each loop:
```
for ( String c : strings ) {
    System.out.println(c);
}
```

It should be possible to exclude for-each loop variables as they might be treated differently from regular local variables.

### PR DESCRIPTION
**Pull request by kris-scheibe**
This PR fixes #1513 

- a new property to exclude the for-each variable (it is "false" by default, so existing rule definitions don't change)
- a test case to check the new property

### CODE REVIEW

**Comment by pmd-test** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="true">This changeset introduces 0 new violations and 0 new errors,<br />
removes 0 violations and 2 errors. <a href="http://chunk.io/pmd/64c2cbbe568940509a5a9d99d3f3d5c4/diff/index.html">Full report</a></td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_danger">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: pmd-java/src/main/resources/category/java/codestyle.xml [hunk 1]
```diff
@@ -1127,6 +1127,9 @@ public interface MissingProperSuffix extends javax.ejb.EJBLocalObject {}    // n
 A local variable assigned only once can be declared final.
         </description>
         <priority>3</priority>
+        <properties>
```
Comment (adangel) [inline pmd-java/src/main/resources/category/java/codestyle.xml]: For java-based rules such as this one, we don't need to declare the properties in XML - it's enough, if it is declared in the rule itself by calling `definePropertyDescriptor(...)`.

So, you can remove this change in codestyle.xml

Comment (kris-scheibe) [reply pmd-java/src/main/resources/category/java/codestyle.xml]: Done!

**Comment by kris-scheibe** [comment]
There are a few reasons why I'd like to exclude loop variables here:
- Generally I think that local variables should be treated differently from loop variables, parameters or other variables that have a special role in the block they're defined i. So a rule that applies to local variables shouldn't check loop variables (I'm aware this might be controversial).
- I'd like to treat for and foreach loops the same: In a for loop the control variable can't be declared final, even if it never changes in the loop body, a foreach allows that... So to have both loops look the same neither should be declared final.
- I agree that changing the loop variable in the loop body might be dangerous, but I'd rather have a new "AvoidReassigningLoopVariable" rule that allows a more fine-grained control over what kind of reassignments are allowed in for/foreach loops. I've described my proposal in #1518 -- I hope you don't mind that I'm stealing the name 😄

**Pull request merged**